src/core/cgroup.c

```diff
@@ -2801,8 +2853,16 @@ static int on_cgroup_inotify_event(sd_event_source *s, int fd, uint32_t revents,
                          * because it was queued before the removal. Let's ignore this here safely. */
 
                         u = hashmap_get(m->cgroup_control_inotify_wd_unit, INT_TO_PTR(e->wd));
-                        if (u)
-                                unit_add_to_cgroup_empty_queue(u);
+                        if (u) {
+                                event = cgroup_control_get_last_event(u);
```

Comment:
Good catch on noticing the shared notifications. Furthermore, notifications can be merged not only in "space" but also in time (throttling).
I.e. we don't necessarily process the *last event* rather the last observed state.

My suggestion would be to implement sort of `cgroup_control_get_state` function that'd return flags for individual substates (i.e. populated, frozen). The actions triggered by the notifications shall be both idempotent and invertible, thus we'd always "apply the state" observed at the time of the most recent notification. (I.e. there is already `cgroup_{frozen|thawed}`, the function to deque unit from empty queue should be simple.)

Although I don't think PID 1 induced events can realistically mix up in such a pathological way, the "state observation" seems more robust should the contents of `cgroup.events` change in the future.

Reply:
I've pushed [freezer-reading-events-expanded](https://github.com/Werkov/systemd/tree/freezer-reading-events-expanded) and [freezer-reading-events](https://github.com/Werkov/systemd/tree/freezer-reading-events). The latter is a rebase of the -expanded branch to squash the changes together. I didn't test the behavior.

Wouldn't it be better to use the squashed version (when this PR isn't merged yet), so that it makes backports more readable?

Reply:
Let me know once this is resolved and I'll be happy to review again :-)